### PR TITLE
Wip mds misc

### PR DESCRIPTION
**Pull request by ukernel**

### CODE REVIEW

**Comment by gregsfortytwo** [comment]
> mds: auth pin CDir when marking it dirty

Why do we want to auth pin a CDir when it's dirty? That will prevent us from migrating it unless the data gets flushed out to the backing objects...don't we go to quite a lot of trouble to migrate dirty data over to the new MDS?

**Comment by ukernel** [comment]
On Tue, Apr 14, 2015 at 1:01 PM, Gregory Farnum <notifications@github.com>
wrote:

> mds: auth pin CDir when marking it dirty
>
> Why do we want to auth pin a CDir when it's dirty? That will prevent us
> from migrating it unless the data gets flushed out to the backing
> objects...don't we go to quite a lot of trouble to migrate dirty data over
> to the new MDS?
>
>
that commit's title is misleading. CDir::log_mark_dirty() not only marks
CDir dirty, but also submits an async context.

**Comment by ukernel** [comment]
http://pulpito.ceph.com/ubuntu-2015-04-19_22:39:04-fs-wip-yan-fs-testing---basic-multi/

only authentication error and the known mds_full error

**Comment by gregsfortytwo** [comment]
Actually, now I'm a little more confused. "mds: auth pin CDir when marking it dirty" doesn't actually include any auth pin calls. It makes an existing auth_unpin() in CDir::_omap_fetched() conditional on whether calling log_mark_dirty...so a more accurate title would be "mds: keep auth pin on CDir when marking it dirty"?

And the reason we need to keep it is so that we can do that later mark dirty, but this case only happens when reading off disk and we have stale snaps?

**Comment by gregsfortytwo** [comment]
"mds: fix Locker::_do_null_snapflush" is dropping a snapshot follows check — is that just a bad copy from other snap handling code?
In general we want the follows to make sure we're keeping all the relevant copies of things, and I'm not familiar enough with the null snapflush code here to know why it's different. Please explain a little more?

**Comment by gregsfortytwo** [comment]
"client: flush dirty snap data before allowing new writes" is just about fixing that waitfor_caps versus waitfor_flush distinction? Was that causing a hang (because caps were already done, but we were on that list and not getting woken up), or something else?

File: src/client/Inode.h [hunk 1]
```diff
@@ -54,6 +54,9 @@ struct CapSnap {
   map<string,bufferptr> xattrs;
   version_t xattr_version;
 
+  bufferlist inline_data;
+  version_t inline_version;
```
Comment (gregsfortytwo) [inline src/client/Inode.h]: You've added these members, but nothing writes them, they're only read.

Comment (ukernel) [inline src/client/Inode.h]: they are modified in Client::finish_cap_snap

**Comment by ukernel** [comment]
On Tue, Apr 21, 2015 at 6:08 AM, Gregory Farnum <notifications@github.com>
wrote:

> "client: flush dirty snap data before allowing new writes" is just about
> fixing that waitfor_caps versus waitfor_flush distinction? Was that causing
> a hang (because caps were already done, but we were on that list and not
> getting woken up), or something else?
>
> --
> Reply to this email directly or view it on GitHub
> <https://github.com/ceph/ceph/pull/4331#issuecomment-94580767>.
>

It's about to avoid new writes overwrite the dirty snapshot data. (because
objectcacher  always allow new data overwrite exiting data) I don't
understand why you think there is a potential hang.

**Comment by ukernel** [comment]
On Tue, Apr 21, 2015 at 5:27 AM, Gregory Farnum <notifications@github.com>
wrote:

> "mds: fix Locker::_do_null_snapflush" is dropping a snapshot follows check
> -- is that just a bad copy from other snap handling code?
> In general we want the follows to make sure we're keeping all the relevant
> copies of things, and I'm not familiar enough with the null snapflush code
> here to know why it's different. Please explain a little more?
>

Locker::_do_null_snapflush() does not trigger any COW, it's used to cleanup
CInode::client_need_snapflush (we add items to
CInode::client_need_snapflush when COWing inode).

Locker::_do_null_snapflush() is called when client releases
CEPH_CAP_ANY_FILE_WR. In most cases, follows in MClientCaps is 0 because
the MClientCaps message does not flush anything)

**Comment by ukernel** [comment]
no change beside update first commit's title

**Comment by dachary** [comment]
@ukernel please excuse the bot spam, I'm having troubles controlling it right now but I'll prevail ;-)

**Pull request merged**